File: myExpenses/src/main/java/org/totschnig/myexpenses/widget/AbstractWidget.java
package org.totschnig.myexpenses.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.appwidget.AppWidgetProviderInfo;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.graphics.Color;
import android.os.Build;
import android.support.annotation.DimenRes;
import android.util.TypedValue;
import android.widget.RemoteViews;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.model.Model;
import org.totschnig.myexpenses.preference.PrefKey;
import org.totschnig.myexpenses.util.UiUtils;

import static android.appwidget.AppWidgetManager.INVALID_APPWIDGET_ID;

public abstract class AbstractWidget<T extends Model> extends AppWidgetProvider {

  public AbstractWidget() {
    super();
  }

  abstract String getPrefName();

  abstract PrefKey getProtectionKey();

  abstract T getObject(Cursor c);

  abstract Cursor getCursor(Context c);

  protected RemoteViews updateWidgetFrom(Context context, int widgetId, int layoutId, T o) {
    RemoteViews updateViews = new RemoteViews(context.getPackageName(), layoutId);
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
      float dimension = context.getResources().getDimension(getFontSizeResId());
      updateViews.setTextViewTextSize(R.id.line1, TypedValue.COMPLEX_UNIT_PX, dimension);
      updateViews.setTextViewTextSize(R.id.note, TypedValue.COMPLEX_UNIT_PX, dimension);
    }
    return updateViews;
  }

  protected static final String WIDGET_NEXT_ACTION = "org.totschnig.myexpenses.UPDATE_WIDGET_NEXT";
  protected static final String WIDGET_PREVIOUS_ACTION = "org.totschnig.myexpenses.UPDATE_WIDGET_PREVIOUS";
  public static final String EXTRA_START_FROM_WIDGET = "startFromWidget";
  public static final String EXTRA_START_FROM_WIDGET_DATA_ENTRY = "startFromWidgetDataEntry";
  protected static final String WIDGET_ID = "widgetId";
  protected static final String PREF_PREFIX_KEY = "prefix_";

  public static void updateWidgets(Context context, Class<? extends AbstractWidget<?>> provider) {
    Intent i = new Intent(context, provider);
    i.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
    int[] widgetIds = AppWidgetManager.getInstance(context).getAppWidgetIds(new ComponentName(context, provider));
    i.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, widgetIds);
    context.sendBroadcast(i);
  }

  protected long loadForWidget(Context context, int widgetId) {
    SharedPreferences prefs = context.getSharedPreferences(getPrefName(), 0);
    return prefs.getLong(PREF_PREFIX_KEY + widgetId, -1);
  }

  protected void saveForWidget(Context context, int widgetId,
                               long objectId) {
    SharedPreferences.Editor prefs = context
        .getSharedPreferences(getPrefName(), 0).edit();
    prefs.putLong(PREF_PREFIX_KEY + widgetId, objectId);
    prefs.apply();
  }

  protected boolean isProtected() {
    return MyApplication.getInstance().isProtected() &&
        !getProtectionKey().getBoolean(false);
  }

  protected void updateWidgets(Context context, AppWidgetManager manager, int[] appWidgetIds,
                               String action) {
    boolean isProtected = isProtected();
    for (int id : appWidgetIds) {
      AppWidgetProviderInfo appWidgetInfo = manager.getAppWidgetInfo(id);
      if (appWidgetInfo != null) {
        RemoteViews remoteViews;
        if (isProtected) {
          remoteViews = protectedUpdate(context);
        } else {
          int layoutId = appWidgetInfo.initialLayout;
          long objectId = loadForWidget(context, id);
          remoteViews = buildUpdate(context, id, layoutId, objectId, action);
        }
        manager.updateAppWidget(id, remoteViews);
      }
    }
  }

  @Override
  public void onReceive(Context context, Intent intent) {
    String action = intent.getAction();
    if (WIDGET_NEXT_ACTION.equals(action) || WIDGET_PREVIOUS_ACTION.equals(action)) {
      int widgetId = intent.getIntExtra(WIDGET_ID, INVALID_APPWIDGET_ID);
      if (widgetId != INVALID_APPWIDGET_ID) {
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        updateWidgets(context, manager, new int[]{widgetId}, action);
      }
    } else {
      super.onReceive(context, intent);
    }
  }

  @Override
  public void onUpdate(Context context, AppWidgetManager manager, int[] appWidgetIds) {
    updateWidgets(context, manager, appWidgetIds, AppWidgetManager.ACTION_APPWIDGET_UPDATE);
  }


  protected RemoteViews errorUpdate(Context context, String message) {
    RemoteViews updateViews = new RemoteViews(context.getPackageName(), R.layout.widget_no_data);
    updateViews.setTextViewText(R.id.object_info, message);
    updateViews.setTextColor(R.id.object_info, Color.RED);
    return updateViews;
  }

  protected RemoteViews protectedUpdate(Context context) {
    String message = context.getString(R.string.warning_password_protected) + " " +
        context.getString(R.string.warning_widget_disabled);
    return errorUpdate(context, message);
  }

  protected RemoteViews noDataUpdate(Context context) {
    return new RemoteViews(context.getPackageName(), R.layout.widget_no_data);
  }

  RemoteViews buildUpdate(Context context,
                          int widgetId, int layoutId, long objectId, String action) {
    Cursor c = getCursor(context);
    T o;
    try {
      int count = c.getCount();
      if (count > 0) {
        if (count == 1 || objectId == -1) {
          if (c.moveToNext()) {
            o = getObject(c);
            return updateWidgetFrom(context, widgetId, layoutId, o);
          }
        } else {
          boolean found = false;
          while (c.moveToNext()) {
            o = getObject(c);
            if (o.getId() == objectId) {
              found = true;
              if (action.equals(WIDGET_NEXT_ACTION)) {
                continue;
              }
              if (action.equals(WIDGET_PREVIOUS_ACTION) && !c.moveToPrevious()) {
                c.moveToLast();
              }
              o = getObject(c);
              return updateWidgetFrom(context, widgetId, layoutId, o);
            } else {
              if (found) {
                return updateWidgetFrom(context, widgetId, layoutId, o);
              }
            }
          }
          c.moveToFirst();
          o = getObject(c);
          return updateWidgetFrom(context, widgetId, layoutId, o);
        }
      }
      return noDataUpdate(context);
    } finally {
      c.close();
    }
  }

  protected void addScrollOnClick(Context context,
                                  RemoteViews updateViews, int widgetId) {
    Intent intent = new Intent(WIDGET_NEXT_ACTION, null, context,
        getClass());
    intent.putExtra(WIDGET_ID, widgetId);
    PendingIntent pendingIntent = PendingIntent.getBroadcast(context, widgetId,
        intent, PendingIntent.FLAG_UPDATE_CURRENT);
    updateViews.setOnClickPendingIntent(R.id.down_icon, pendingIntent);
    intent = new Intent(WIDGET_PREVIOUS_ACTION, null, context,
        getClass());
    intent.putExtra(WIDGET_ID, widgetId);
    pendingIntent = PendingIntent.getBroadcast(context, widgetId, intent,
        PendingIntent.FLAG_UPDATE_CURRENT);
    updateViews.setOnClickPendingIntent(R.id.up_icon, pendingIntent);
  }

  protected void setBackgroundColorSave(RemoteViews updateViews, int res, int color) {
    updateViews.setInt(res, "setBackgroundColor", color);
  }

  //http://stackoverflow.com/a/35633411/1199911
  protected void setImageViewVectorDrawable(Context context, RemoteViews remoteViews,
                                            int viewId, int resId) {
    remoteViews.setImageViewBitmap(viewId, UiUtils.getTintedBitmapForTheme(context, resId,
        R.style.ThemeDark));
  }

  private @DimenRes int getFontSizeResId() {
    switch(PrefKey.UI_FONTSIZE.getInt(0)) {
      case 3: return R.dimen.textSizeSmallS3;
      case 2: return R.dimen.textSizeSmallS2;
      case 1: return R.dimen.textSizeSmallS1;
      default: return R.dimen.textSizeSmallBase;
    }
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/widget/AbstractWidget.kt
package org.totschnig.myexpenses.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.net.Uri
import android.os.Bundle
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import androidx.appcompat.app.AppCompatDelegate
import androidx.preference.PreferenceFragmentCompat
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.activity.MyPreferenceActivity
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.model.CurrencyContext
import org.totschnig.myexpenses.myApplication
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.preference.PrefKey
import timber.log.Timber
import javax.inject.Inject


const val WIDGET_CLICK = "org.totschnig.myexpenses.WIDGET_CLICK"
const val KEY_CLICK_ACTION = "clickAction"
const val WIDGET_LIST_DATA_CHANGED = "org.totschnig.myexpenses.LIST_DATA_CHANGED"
const val WIDGET_CONTEXT_CHANGED = "org.totschnig.myexpenses.CONTEXT_CHANGED"
const val EXTRA_START_FROM_WIDGET = "startFromWidget"
const val EXTRA_START_FROM_WIDGET_DATA_ENTRY = "startFromWidgetDataEntry"
const val KEY_WIDTH = "width"
const val WIDGET_ROW_RESERVED_SPACE_FOR_INFO = 110

fun onConfigurationChanged(context: Context) {
    updateWidgets(context, AccountWidget::class.java, WIDGET_CONTEXT_CHANGED)
    updateWidgets(context, TemplateWidget::class.java, WIDGET_CONTEXT_CHANGED)
}

fun updateWidgets(
    context: Context, provider: Class<out AppWidgetProvider?>, action: String,
    appWidgetIds: IntArray = AppWidgetManager.getInstance(context)
        .getAppWidgetIds(ComponentName(context, provider))
) =
    context.sendBroadcast(Intent(context, provider).apply {
        this.action = action
        putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, appWidgetIds)
    })

abstract class AbstractWidget(
    private val clazz: Class<out RemoteViewsService>,
    private val protectionKey: PrefKey
) : AppWidgetProvider() {
    abstract val emptyTextResourceId: Int

    @Inject
    lateinit var prefHandler: PrefHandler

    @Inject
    lateinit var currencyContext: CurrencyContext

    override fun onReceive(context: Context, intent: Intent) {
        context.injector.inject(this)
        val instance = AppWidgetManager.getInstance(context)
        val appWidgetIds = intent.extras?.getIntArray(AppWidgetManager.EXTRA_APPWIDGET_IDS)
        when (intent.action) {
            WIDGET_LIST_DATA_CHANGED -> {
                appWidgetIds?.let { instance.notifyAppWidgetViewDataChanged(it, R.id.list) }
            }

            WIDGET_CONTEXT_CHANGED -> {
                appWidgetIds?.let { onUpdate(context, instance, it) }
            }

            WIDGET_CLICK -> {
                handleWidgetClick(context, intent)
            }

            else -> {
                super.onReceive(context, intent)
            }
        }
    }

    abstract fun handleWidgetClick(context: Context, intent: Intent)

    fun availableWidth(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int
    ) = appWidgetManager.getAppWidgetOptions(appWidgetId).getInt(
        when (context.resources.configuration.orientation) {
            Configuration.ORIENTATION_LANDSCAPE -> AppWidgetManager.OPTION_APPWIDGET_MAX_WIDTH
            else -> AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH
        }, Int.MAX_VALUE
    ) - (WIDGET_ROW_RESERVED_SPACE_FOR_INFO * context.resources.configuration.fontScale).toInt()

    fun clickBaseIntent(context: Context) = Intent(WIDGET_CLICK, null, context, javaClass)

    open fun updateWidget(context: Context, appWidgetManager: AppWidgetManager, appWidgetId: Int) {

        appWidgetManager.updateAppWidget(appWidgetId, if (isProtected(context)) {
            RemoteViews(context.packageName, R.layout.widget_locked).apply {
                setOnClickPendingIntent(
                    R.id.text,
                    PendingIntent.getActivity(
                        context,
                        appWidgetId,
                        Intent(context, MyPreferenceActivity::class.java).putExtra(
                            PreferenceFragmentCompat.ARG_PREFERENCE_ROOT,
                            context.getString(R.string.pref_screen_protection_key)
                        ),
                        PendingIntent.FLAG_IMMUTABLE
                    )
                )
                setTextViewText(
                    R.id.text, context.getString(R.string.warning_password_protected) + " " +
                            context.getString(R.string.warning_widget_disabled)
                )
            }
        } else {
            val clickPI = PendingIntent.getBroadcast(
                context,
                appWidgetId,
                clickBaseIntent(context),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
            )

            RemoteViews(
                context.packageName,
                when (AppCompatDelegate.getDefaultNightMode()) {
                    AppCompatDelegate.MODE_NIGHT_NO -> R.layout.widget_list_light
                    AppCompatDelegate.MODE_NIGHT_YES -> R.layout.widget_list_dark
                    else -> R.layout.widget_list
                }
            ).apply {
                setEmptyView(R.id.list, R.id.emptyView)
                setOnClickPendingIntent(R.id.emptyView, clickPI)

                setRemoteAdapter(R.id.list, Intent(context, clazz).apply {
                    putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
                    val availableWidth = availableWidth(context, appWidgetManager, appWidgetId)
                    Timber.i("availableWidth: %d", availableWidth)
                    putExtra(KEY_WIDTH, availableWidth)
                    // When intents are compared, the extras are ignored, so we need to embed the extras
                    // into the data so that the extras will not be ignored.
                    data = Uri.parse(toUri(Intent.URI_INTENT_SCHEME))
                })
                setTextViewText(
                    R.id.emptyView,
                    context.getString(emptyTextResourceId)
                )
                setPendingIntentTemplate(R.id.list, clickPI)
            }
        })
    }

    override fun onAppWidgetOptionsChanged(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int,
        newOptions: Bundle
    ) {
        updateWidget(context, appWidgetManager, appWidgetId)
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        appWidgetIds.forEach { appWidgetId ->
            updateWidget(context, appWidgetManager, appWidgetId)
        }
    }

    protected open fun isProtected(context: Context): Boolean {
        return (context.myApplication).isProtected &&
                !prefHandler.getBoolean(protectionKey, false)
    }
}